source/agora/common/Deserializer.d

```diff
@@ -188,35 +193,113 @@ public void deserializePart (ref ubyte record, scope DeserializeDg dg)
 public void deserializePart (ref ushort record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(ushort*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
 public void deserializePart (ref uint record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(uint*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
-public void deserializePart (ref ulong record, scope DeserializeDg dg)
+public void deserializePart (ref ulong record, scope DeserializeDg dg, CompactMode compact = CompactMode.Yes)
     @trusted
 {
-    record = *cast(ulong*)(dg(record.sizeof).ptr);
+    if (compact == CompactMode.Yes)
+        deserializeVarInt(record, dg);
+    else
+        record = *cast(ulong*)(dg(record.sizeof).ptr);
 }
 
 /// Ditto
 public void deserializePart (ref char[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = cast(char[])dg(length);
 }
 
 /// Ditto
 public void deserializePart (ref ubyte[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = dg(length);
 }
+
+/*******************************************************************************
+
+    Deserialize varint a struct and convert T type return it.
+
+    VarInt Size
+    size_tag is first a ubyte
+    size_tag <= 253(0xFC)  -- 1 byte   ubyte
+    size_tag == 0xFD       -- 3 bytes  (0xFD + ushort)
+    size_tag == 0xFE       -- 5 bytes  (0xFE + uint)
+    size_tag == 0xFF       -- 9 bytes  (0xFF + ulong)
+
+    Params:
+        T = Type of unsigned integer to deserialize
+        num = Binary serialized representation of `T`
+                 to be deserialized varint
+        dg = source of binary data
+
+*******************************************************************************/
+
+public void deserializeVarInt (T) (ref T num, scope DeserializeDg dg)
+    @trusted
+    if (is(T == ushort) || is(T == uint) || is(T == ulong))
+{
+    const ubyte int_size = dg(ubyte.sizeof)[0];
+    assert(int_size >= 0);
+
+    T read (InType)()
```

Comment:
Looks like you forgot to actually use the function?

Reply:
T does not support ubyte here.
I wonder if this passage is necessary.

Reply:
That's why I didn't use it for ubytes. The code I showed was this:

```D
public void deserializeVarInt (T) (ref T record, scope DeserializeDg dg)
    @trusted
    if (is(T == ushort) || is(T == uint) || is(T == ulong))
{
    const ubyte int_size = dg(ubyte.sizeof)[0];
    assert(int_size >= 0);

    T read (InType)()
    {
        import std.exception;
        auto value = *cast(InType*)dg(InType.sizeof).ptr;
        static if (T.max < InType.max)
            enforce(value <= T.max);
        return cast(T)value;
    }

    if (int_size <= 0xFC)
        record = cast(T)(int_size);
    else if (int_size == 0xFD)
        record = read!ushort();
    else if (int_size == 0xFE)
        record = read!uint();
    else if (int_size == 0xFF)
        record = read!ulong();
    else
        assert(0);
}
```

Reply:
Oh, I missed it. Thank you.